commit ee6a569d3bf64c9676eee3eecb861fb01cc11311
Author: Michael Mueller <mimu@linux.ibm.com>
Date:   Wed Feb 9 16:22:17 2022 +0100

    KVM: s390: pv: make use of ultravisor AIV support
    
    This patch enables the ultravisor adapter interruption vitualization
    support indicated by UV feature BIT_UV_FEAT_AIV. This allows ISC
    interruption injection directly into the GISA IPM for PV kvm guests.
    
    Hardware that does not support this feature will continue to use the
    UV interruption interception method to deliver ISC interruptions to
    PV kvm guests. For this purpose, the ECA_AIV bit for all guest cpus
    will be cleared and the GISA will be disabled during PV CPU setup.
    
    In addition a check in __inject_io() has been removed. That reduces the
    required instructions for interruption handling for PV and traditional
    kvm guests.
    
    Signed-off-by: Michael Mueller <mimu@linux.ibm.com>
    Reviewed-by: Janosch Frank <frankja@linux.ibm.com>
    Link: https://lore.kernel.org/r/20220209152217.1793281-2-mimu@linux.ibm.com
    Reviewed-by: Christian Borntraeger <borntraeger@linux.ibm.com>
    Signed-off-by: Christian Borntraeger <borntraeger@linux.ibm.com>

diff --git a/arch/s390/kvm/kvm-s390.h b/arch/s390/kvm/kvm-s390.h
index 098831e815e6..4ba8fc30d87a 100644
--- a/arch/s390/kvm/kvm-s390.h
+++ b/arch/s390/kvm/kvm-s390.h
@@ -231,6 +231,15 @@ static inline unsigned long kvm_s390_get_gfn_end(struct kvm_memslots *slots)
 	return ms->base_gfn + ms->npages;
 }
 
+static inline u32 kvm_s390_get_gisa_desc(struct kvm *kvm)
+{
+	u32 gd = (u32)(u64)kvm->arch.gisa_int.origin;
+
+	if (gd && sclp.has_gisaf)
+		gd |= GISA_FORMAT1;
+	return gd;
+}
+
 /* implemented in pv.c */
 int kvm_s390_pv_destroy_cpu(struct kvm_vcpu *vcpu, u16 *rc, u16 *rrc);
 int kvm_s390_pv_create_cpu(struct kvm_vcpu *vcpu, u16 *rc, u16 *rrc);
@@ -450,6 +459,8 @@ int kvm_s390_get_irq_state(struct kvm_vcpu *vcpu,
 void kvm_s390_gisa_init(struct kvm *kvm);
 void kvm_s390_gisa_clear(struct kvm *kvm);
 void kvm_s390_gisa_destroy(struct kvm *kvm);
+void kvm_s390_gisa_disable(struct kvm *kvm);
+void kvm_s390_gisa_enable(struct kvm *kvm);
 int kvm_s390_gib_init(u8 nisc);
 void kvm_s390_gib_destroy(void);